### ISSUE TITLE
Reproduce Reconstruction Results on the GSO dataset

### ISSUE THREAD

**Issue by lzqsd** [TencentARC/FreeSplatter#8]
Thanks for open source this great work to community!

I was running the inference demo on my local machine using images rendered from the GSO dataset. However, I cannot reproduce the results shown in the paper. Followings are 2 example results I obtained from the interactive demo. Would you please tell me if there is anything i should pay attention to? I can also send the testing images to you directly. Really appreciate your help and guidance in advance!

![Image](https://github.com/user-attachments/assets/eae2ec4d-e5e0-4541-b9ae-26e79d76e709)

![Image](https://github.com/user-attachments/assets/08593b01-f12c-40a4-9709-4f9b782b229b)

**Comment by ncan33** [comment]
I am also having a really hard time reproducing the results on the paper. The paper has Figure 1 like this:

<img width="1350" height="511" alt="Image" src="https://github.com/user-attachments/assets/7c95fbee-a13f-4f2e-9c29-c708d31cd304" />

The images for this figure are provided by the authors in the "examples" folder in the [HuggingFace repository](https://huggingface.co/spaces/TencentARC/FreeSplatter/tree/main):

<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/6a52100e-76a6-4fb9-9e93-19bfa0efae3f" />
<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/945e6c2a-2398-4a85-bb68-3e38dbd12301" />
<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/2c9145c6-8569-4547-b693-af8e8359b439" />
<img width="128" height="128" alt="Image" src="https://github.com/user-attachments/assets/d5532db0-0f41-4434-bc61-9cffa52794d9" />

Using these images, I called run_views_to_3d:

```
import torch
import glob
import os
from freesplatter.webui.runner import FreeSplatterRunner

if torch.cuda.is_available():
    runner = FreeSplatterRunner(device="cuda")
elif not torch.cuda.is_available():
    raise ExceptionType("Do not even try running this without a GPU.")

image_paths = sorted(glob.glob(os.path.join("gt_test", "images/*.png")))

images_vis, gs_vis_path, video_path, mesh_path, pose_fig = runner.run_views_to_3d(
    image_files=image_paths,
    do_rembg=False,
    gs_type='2DGS',
    cache_dir="gt_test"
)
```

But with these images, this is the reconstruction I get:

https://github.com/user-attachments/assets/a51cdf38-1839-43bf-aecb-092da1a34077

There's a big difference between this result and what we are seeing in the paper. Not sure how authors produced Figure 1. I am wondering if they took the pose estimates (which are created from plugging predicted Gaussians into the PnP-RANSAC solver) and plugged it into a conventional 3DGS reconstruction? The reported reconstructions are super quality but the predicted Gaussians I get from `run_views_to_3d()` and `run_freesplatter_object()` do not match the results I am seeing in the paper.

Authors, am I doing something wrong here? I cannot reproduce the results in the paper.

**Comment by bluestyle97** [comment]
@ncan33 The results you get look really weird. Did you initialize the environment following the guidance of this repo?

**Comment by ncan33** [comment]
@bluestyle97 I did set up the environment following the guidance of the repo. All of the dependencies installed and work well. I ran some experiments this morning. For the diffusion-based reconstruction, I am getting really strong results using `run_img_to_3d()`. For example, for this input image:

<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/2f626475-560d-4603-9787-6ecdf5cb470d" />

I get this diffusion synthesis:

<img width="1920" height="320" alt="Image" src="https://github.com/user-attachments/assets/f140ce2c-0238-45d8-aa30-871f67842e9c" />

And the reconstruction looks great:

https://github.com/user-attachments/assets/bcf0f46c-a63f-4f11-9bbe-f46fc518dd5e

However, when I reconstruct using the **real images of the object taken from multiple angles** using `run_views_to_3d()`, the reconstruction looks bad:

https://github.com/user-attachments/assets/0bae0f25-a1e3-4144-babc-daaa528ad4f9

The **real images of the object taken from multiple angles** are:

<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/d395e51e-6965-43ef-8902-bd0f3e7fdb77" />
<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/0109cdc7-2278-4915-8610-8238dfc41b23" />
<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/aee5f640-e83f-48b6-b299-aab48ffdd3e9" />
<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/5487d318-cae7-493e-a5a5-7810728772cd" />
<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/803aa156-0617-4867-9511-3e36942a4ca2" />

I would expect that the reconstruction with the real images would outperform the reconstruction with the diffusion images (except for the bottom of the object maybe). However, it seems like the single-image reconstruction is outperforming the multi-image reconstruction.

I might be doing something wrong with the data preparation for `run_views_to_3d()`. To isolate the issue, I will take the diffusion images generated from `run_img_to_3d()` and plug input them into `run_views_to_3d()`, and see if I get the same result as the first video above (gs.12.mp4), or if it will look like the second video (gs.10.mp4).

I think FreeSplatter is an excellent, novel contribution. I think I am probably doing something wrong with the data prep. I will investigate more and submit and update with my findings.

**Comment by ncan33** [comment]
I think there were some inconsistencies in the way that `run_views_to_3d()` does data prep vs. the way `run_img_to_3d()` does data prep. I copied the data prep logic from `run_img_to_3d()` and I got this reconstruction with the **real images of the object taken from multiple angles**:

https://github.com/user-attachments/assets/05355c68-1842-4c33-996e-dc4c788201d3

This seems more reasonable. It's still a little more fuzzy/cloudy than the diffusion one, but that might be attributed to differences in viewing angles maybe. Not sure. For a more scientific comparison, I should capture photos of this mint box from the same exact viewing angles as how diffusion synthesized it.

After this fix, I re-ran the "Transformers" model and this is what I got:

https://github.com/user-attachments/assets/e09c95d7-79b7-4d57-b3c7-edc518b39622

This is substantially better than before. Still a little cloudy, especially on the back. I think I still need to fix some stuff and make it better match the data prep of `run_img_to_3d()`

Thanks for this excellent paper! I will give another update if I can push the reconstruction quality.

**Comment by bluestyle97** [comment]
@ncan33 FreeSplatter-O was trained on Objaverse-rendered images. Each model was normalized into [-1, 1]^3 unit cube, and cameras are randomly sampled around the object, **all facing the bounding box center [0, 0, 0]**. This data rendering strategy follows the typical practice of previous LRMs, but it indeed leads to a camera distribution mis-matching problem when using real-captured images for inference, whose cameras often do not share a common "center". That's why you got better results using diffusion-generated images than real photos, since the multi-view diffusion model (Hunyuan3D or Zero123++) was also trained with centered cameras, making their generated views conform to the camera distribution towards the center. To improve the results, you can refer to a follow-up work named [UAV-Splatter](https://github.com/yfujimura/UFV-Splatter), which alleviate this issue by adding lora layers to FreeSplatter and fine-tuning with more diverse image.
Besides, the results you got with the example "Transformer" images look still significantly worse than what we got. I have no idea on the exact reason for now. Maybe I can look into this issue this weekend. Here is our reconstruction results for reference: [Gaussians](https://huggingface.co/datasets/bluestyle97/FreeSplatterStatic/blob/main/webpage_object/GSO_Transformers_Age_of_Extinction_Mega_1Step_Bumblebee_Figure_gs.ply) & [video](https://huggingface.co/datasets/bluestyle97/FreeSplatterStatic/blob/main/webpage_object/GSO_Transformers_Age_of_Extinction_Mega_1Step_Bumblebee_Figure_video.mp4).
Thanks for your interest on our work and the valuable explorations!

**Comment by ncan33** [comment]
Thank you for this clear explanation and sharing [UAV-Splatter](https://github.com/yfujimura/UFV-Splatter). Thank you also for sharing the reconstructed Gaussians and video. Excited to follow your future works!

**Comment by ncan33** [comment]
Follow up to this: I was able to exactly reproduce the results of this paper on a variety of images. For example, here are my results on the "Transformers" model. There is no cloudiness whatsoever, and it is identical the the [results of the authors](https://huggingface.co/datasets/bluestyle97/FreeSplatterStatic/blob/main/webpage_object/GSO_Transformers_Age_of_Extinction_Mega_1Step_Bumblebee_Figure_video.mp4). My results:

https://github.com/user-attachments/assets/fe772f35-d6f7-467e-8360-c5ece7c929ca

**Comment by kk627627** [comment]
@ncan33 

> Follow up to this: I was able to exactly reproduce the results of this paper on a variety of images. For example, here are my results on the "Transformers" model. There is no cloudiness whatsoever, and it is identical the the [results of the authors](https://huggingface.co/datasets/bluestyle97/FreeSplatterStatic/blob/main/webpage_object/GSO_Transformers_Age_of_Extinction_Mega_1Step_Bumblebee_Figure_video.mp4). My results:
> 
>  gs.mp4

What specific strategies and improvements did you implement to achieve good results when inputting multi-view images?
